### ISSUE TITLE
Intersection type with partial not working

### ISSUE THREAD

**Issue by jvanbruegge** [microsoft/TypeScript#16251]
**TypeScript Version:**  2.2.1

**Code**

```ts
export type Transformer<Si> = (name: string) => (s: Array<Instances<Si>>) => Array<any>;
export type TransformerConfig<Si> = { [k in keyof Si]?: Transformer<Si> };
//export type TransformerConfig<Si> = Partial<{ [k in keyof Si]: Transformer<Si> }>; <-- same result
export type ToSinksConfig<Si> = { '*': Transformer<Si> } & TransformerConfig<Si>;
//export type ToSinksConfig<Si> = { '*': Transformer<Si> } & TransformerConfig<Si> | { '*': Transformer<Si> }; <-- works
export type Instances<Si> = { ins: Array<Si> };

class Configurator<Si> {
    public configure(config: ToSinksConfig<Si> = { '*': n => x => x }): any {
        return undefined;
    }
}
```

**Expected behavior:**
No errors

**Actual behavior:**
```
test.ts(7,22): error TS2322: Type '{ '*': (n: string) => (x: Instances<Si>[]) => Instances<Si>[]; }' is not assignable to type 'ToSinksConfig<Si>'.
  Type '{ '*': (n: string) => (x: Instances<Si>[]) => Instances<Si>[]; }' is not assignable to type 'TransformerConfig<Si>'.
```

**Comment by ikatyang** [comment]
For minimal reproduced code and some information:

v2.4.0-dev.20170604

```ts
type X1<T> = { x: number } & { [K in keyof T]?: any };
function fn1<T>(v: X1<T> = { x: 1 }): any {};
// unexpected error
// Type '{ x: number; }' is not assignable to type 'X1<T>'.
//   Type '{ x: number; }' is not assignable to type '{ [K in keyof T]?: any; }'.

type X2<T> = { x: number } & { [K in keyof any]?: any };
function fn2<T>(v: X2<T> = { x: 1 }): any {};
// no error as expected

type X3<T> = { x: number } & { [K in string]?: any };
function fn3<T>(v: X3<T> = { x: 1 }): any {};
// no error as expected
```

**Comment by tycho01** [comment]
Still fails with `type X1<T extends string> = { x: number } & { [K in keyof T]?: any };` (make `T` explicitly extend `string`). Huh.

**Comment by ashmind** [comment]
Same problem when using `Partial<T>` I think:
```typescript
function f<T>() {
    // Type '{ a: 1; }' is not assignable to type '{ a: 1; } & Partial<T>'.
    //  Type '{ a: 1; }' is not assignable to type 'Partial<T>'
    const x: { a: 1 } & Partial<T> = { a: 1 };
}
```

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMAPAFQHwAoCUiDeAUIiYhAgM5SIAeAXAYgIYMCMiAvogGSIAKTAE6wmAG3QZEAXkYtE7DgG5CHIA

**Comment by jakebailey** [comment]
Just checking out old issues; this looks to have been fixed by #32071.

For the code in the recent comment, the error looks correct to me; T could be `{ a: string }` in which case you're assigning to `never` (or, just plain the fact that it could be undefined and it can't be `1 & undefined`.
